quickwit-indexing/src/garbage_collection.rs

```diff
@@ -0,0 +1,175 @@
+        .list_splits(index_id, SplitState::Staged, None, &[])
+        .await?
+        .into_iter()
+        // TODO: Update metastore API and push this filter down.
+        .filter(|meta| meta.split_metadata.update_timestamp < grace_period_timestamp)
+        .collect();
+
+    if dry_run {
+        let mut scheduled_for_delete_splits = metastore
+            .list_splits(index_id, SplitState::ScheduledForDeletion, None, &[])
+            .await?;
+        scheduled_for_delete_splits.extend(deletable_staged_splits);
+
+        let candidate_entries: Vec<FileEntry> = scheduled_for_delete_splits
+            .iter()
+            .map(FileEntry::from)
+            .collect();
+        return Ok(SplitDeletionStats {
+            candidate_entries,
+            ..Default::default()
+        });
+    }
+
+    // Schedule all eligible staged splits for delete
+    let split_ids: Vec<&str> = deletable_staged_splits
+        .iter()
+        .map(|meta| meta.split_metadata.split_id.as_str())
+        .collect();
+    metastore
+        .mark_splits_as_deleted(index_id, &split_ids)
+        .await?;
+
+    // Select split to delete
+    let splits_to_delete = metastore
+        .list_splits(index_id, SplitState::ScheduledForDeletion, None, &[])
+        .await?;
+    let deleted_files = delete_splits_with_files(
+        index_id,
+        storage.clone(),
+        metastore.clone(),
+        splits_to_delete,
+    )
+    .await?;
+
+    Ok(deleted_files)
+}
+
+/// Delete a list of splits from the storage and the metastore.
+/// It should leave the index and the metastore in good state.
+///
+/// * `index_id` - The target index id.
+/// * `storage - The storage managing the target index.
+/// * `metastore` - The metastore managing the target index.
+/// * `splits`  - The list of splits to delete.
+pub async fn delete_splits_with_files(
+    index_id: &str,
+    storage: Arc<dyn Storage>,
+    metastore: Arc<dyn Metastore>,
+    splits: Vec<SplitMetadataAndFooterOffsets>,
+) -> anyhow::Result<SplitDeletionStats> {
+    let mut deletion_stats = SplitDeletionStats::default();
+    let mut deleted_split_ids: Vec<String> = Vec::new();
+    let mut failed_split_ids: Vec<String> = Vec::new();
+
+    let mut delete_splits_results_stream = tokio_stream::iter(splits.into_iter())
+        .map(|meta| {
+            let moved_storage = storage.clone();
+            async move {
+                let file_entry = FileEntry::from(&meta);
+                let delete_result = moved_storage.delete(Path::new(&file_entry.file_name)).await;
+                (
+                    meta.split_metadata.split_id.clone(),
+                    file_entry,
+                    delete_result.is_ok(),
+                )
+            }
+        })
+        .buffer_unordered(MAX_CONCURRENT_STORAGE_REQUESTS);
+
+    while let Some((split_id, file_entry, is_success)) = delete_splits_results_stream.next().await {
```

Comment:
just a personal nitpick: I would use `.collect()`, log when there is an error and filter out errors and ends with just a vec of `split_id`.